recipes/qt/all/conanfile.py

```diff
@@ -0,0 +1,673 @@
+import os
+import shutil
+import itertools
+import glob
+
+import configparser
+from conans import ConanFile, tools, __version__ as conan_version, RunEnvironment
+from conans.errors import ConanInvalidConfiguration
+from conans.model import Generator
+from conans.tools import Version
+
+
+class qt(Generator):
+    @property
+    def filename(self):
+        return "qt.conf"
+
+    @property
+    def content(self):
+        return "[Paths]\nPrefix = %s\n" % self.conanfile.deps_cpp_info["qt"].rootpath.replace("\\", "/")
+
+
+def _getsubmodules():
+    config = configparser.ConfigParser()
+    config.read('qtmodules.conf')
+    res = {}
+    assert config.sections()
+    for s in config.sections():
+        section = str(s)
+        assert section.startswith("submodule ")
+        assert section.count('"') == 2
+        modulename = section[section.find('"') + 1: section.rfind('"')]
+        status = str(config.get(section, "status"))
+        if status != "obsolete" and status != "ignore":
+            res[modulename] = {"status": status,
+                               "path": str(config.get(section, "path")), "depends": []}
+            if config.has_option(section, "depends"):
+                res[modulename]["depends"] = [str(i) for i in config.get(section, "depends").split()]
+    return res
+
+
+class QtConan(ConanFile):
+    _submodules = _getsubmodules()
+
+    generators = "pkg_config"
+    name = "qt"
+    description = "Qt is a cross-platform framework for graphical user interfaces."
+    topics = ("conan", "qt", "ui")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.qt.io"
+    license = "LGPL-3.0"
+    exports = ["qtmodules.conf", "patches/*.diff"]
+    settings = "os", "arch", "compiler", "build_type"
```

Comment:
At least on linux, bsd, android, macos, ios, qnx, you can use a compiler to build Qt and another one to build your apps.
Adding the "compiler" to setting to all platforms will force a qt build most of the time, which usually takes a lot!
Is it possible to remove "compiler" from settings for these platforms?

Reply:
the compiler has a direct impact on the compiled binaries, so it does not seem to be a good move. Why do you consider qt different from other libraries in this regard ?

Reply:
What makes you think that The Qt Company builds Qt SDK for each linux distro and compiler brand and compiler version out there? Same for linux distros, do you think that debian, will rebuild all their packages each time when they add a new compiler :) ?
I agree that the compiler has a direct impact on the bin itself (i.e. the quality of the code), but it should not have any impact on the libs ABIs (unless you're using super compiler specific flags which breaks the compatibility with other compilers).
On LInux you can build a library with gcc 5, then build your application which uses that library using gcc 10 without any problems. Also the vice-versa works as long as the library is not using newer std libs (libc, libsstdc++, etc.). You can even use different compilers e.g. build a library A with gcc, library B with clang and your application (which uses both libs) with Intel C compiler.

On my linux setup I have 3 gcc versions and 2 clang, and I can use all of them with all my system libs.

Besides MSVC which really needs to use the same version, I have no idea why conan folks chose to enforce the compiler brand and version ...

Reply:
> I have no idea why conan folks chose to enforce the compiler brand and version ...

I am not a Conan guy, but I guess it is a cheap/defensive way of ensuring ABI compatibility without calling it ABI. See also https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html. 

Obviously, a combination of object file format (ELF/PE/Mach-o), architecture, ABI and the std libs version would probably be more precise.

Reply:
even if you're not changing flags for different compilers, compilers themselves use different default flags for various versions.
for instance, if we check GCC 5:
```
conan@9b7367953ef9:~$ g++ -x c++ -E -dM /dev/null | grep cplus
#define __cplusplus 199711L
```
then check GCC 10:
```
conan@b1f079ddd464:~$ g++ -x c++ -E -dM /dev/null | grep cplus
#define __cplusplus 201402L
``` 
what does it mean? GCC 5 will use **c++98** standard by default, why GCC 10 will use **c++14** by default.

now imagine we have the following code:
```
#if __cplusplus >= 201103L
std::unique_ptr<data> get_data();
#else
std::auto_ptr<data> get_data();
#endif
```
this is a typical C++ API header which needs to support several standard, and it has to met the following conditions:
- auto_ptr was removed in C++17, and available only in older standards
- unique_ptr was introduced in C++11 only

therefore, either way the code must contain conditional compilation. if you think it's synthetic case - it's very real, and boost has very similar branch in API (it exports different symbols for different C++ standards). and boost is not alone here, other libraries do the same.

so, if you try to mix libraries build with GCC 5 and GCC 10 - you will either get linking errors or runtime errors (much worse).

C++ standard is just one of many possible things. for instance, if GCC 10 added support for more intrinsics, and library managed to conditionally use them if available, linking library into application compiled with GCC will also probably fail.

of course, on some simple libraries with plain C API it's hard to manifest such things, and you will probably never run into the problems, but in generic case their are not so compatible as it appears on the first view.

Reply:
Of course there are some libraries that breaks the ABI with different compilers, but Qt 5.15 is not the case.

As I said it depends on libs which are using incompatible ABIs, but if you'll build boost with a C++11 enabled compiler (e.g. gcc 5) it should work just fine with gcc 10 (even with c++20 turned on). Obviously I don't expect that boost built with c++20 to work on a compiler that supports only c++11.

Reply:
as I have said, default flags may change between compiler releases, and C++ standard is only one example of such change.
and boost built with C++11 compiler will not be compatible with boost built with C++20 compiler (as auto_ptr removed in C++17, as well as few other things removed in C++17 and C++20), so it's not a safe to mix different C++ standards in general.
anyway, if you're absolutely sure that your particular library (e.g. Qt) is compatible across a range of different compilers and their versions, you can easily model that with `package_id` method. that's just not something to assume in general by default.

Reply:
The problem is that nobody is using `package_id`, not even conan folks, e.g. https://github.com/conan-io/conan-center-index/blob/master/recipes/libpng, libssh2, libtiff, libx264 :) All of these al plain C libs which they don't need to be restricted to a compiler type or version on unix systems (yes, for msvc  it's a must).

I think that most of the libs are compatible across different compiler types and versions and just a few are not, so, IMHO the compiler should not be automatically added to the `package_id` (unless it's msvc).
If a package has a problem and a compiler type/version then the user can easily rebuild it with his own compiler. The problem is that the user can't download pre-built bins if the compiler type and version don't match perfectly even the library will work just fine which his compiler...